File: saler/src/main/java/com/saler/saler/deal/service/DealService.java
package com.saler.saler.deal.service;

import java.util.List;

import com.saler.saler.buy.option.entity.BuyOption;
import com.saler.saler.deal.entity.Deal;

/**
 * Deal service
 * @author Diego Andre Poli
 */
public interface DealService {

	/**
	 * Get deal by id
	 * @param id of deal
	 * @return deal
	 */
	Deal getById(Long id);
	
	/**
	 * Get all deals
	 * @return list of deals
	 */
	List<Deal> getAll();
	
	/**
	 * Add deal
	 * @param deal
	 * @return deal added
	 */
	Deal add(Deal deal);
	
	/**
	 * Remove deal
	 * @param id of deal
	 */
	void remove(Long id);
	
	/**
	 * Get id by url
	 * @param deal url
	 * @return deal id
	 */
	Long getIdByUrl(String url);
	
	/**
	 * Add buy option 
	 * @param deal id 
	 * @param buy option
	 * @return deal
	 */
	Deal addBuyOption(Long id, BuyOption buyOption);
	
	/**
	 * Decrement cupon
	 * @param deal id
	 */
	void decrementCupon(Long id);
	
	/**
	 * Get by url
	 * @param deal url
	 * @return deal
	 */
	Deal getByUrl(String url);
	
	/**
	 * Save or update deal
	 * @param deal
	 * @return deal
	 */
	Deal saveOrUpdate(Deal deal);
}

File: midle/src/app/deal.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Configuration, Utils } from './app.constants';


/**
 * Deal service
 * @author Diego Andre Poli <diegoandrepoli@gmail.com>
 */
@Injectable()
export class DealService {

  /**
   * Server URL
   */
  private actionUrl: string;

  /**
   * Custom construcotor
   * @param http
   */
  constructor(private http: HttpClient, private _utils: Utils, private _configuration: Configuration) {
    this.actionUrl = _configuration.server;
    let headers = this._utils.getHeaders();
  }

  /**
   * Get all deals
   * @return list of deals
   */
  public getAll<T>(): Observable<T> {
      return this.http.get<T>(this.actionUrl.concat('/deal'));
  }

  /**
   * Add deal
   * @param deal
   */
  public add<T>( item: Object ): Observable<T> {
      return this.http.post<T>( this.actionUrl.concat('/deal/add'), item);
  }

  /**
   * Update deal
   * @param deal
   */
  public update<T>( item: Object ): Observable<T> {
    return this.http.put<T>( this.actionUrl.concat('/deal'), item);
  }

  /**
   * Delete deal
   * @param id of deal
   */
  public delete<T>( id: string ): Observable<T> {
    return this.http.delete<T>( this.actionUrl.concat('/deal/').concat(id));
  }

  /**
   * Get by id
   * @param id of deal
   */
  public getById<T>( id: string ): Observable<T> {
    return this.http.get<T>( this.actionUrl.concat('/deal/').concat(id));
  }

  /**
   * Get by url
   * @param url as deal
   */
  public getByUrl<T>( url: string ): Observable<T> {
    return this.http.get<T>( this.actionUrl.concat('/deal/url/').concat(url));
  }

  /**
   * Byu!
   * @param deal id
   * @param buy option id
   */
  public buy<T>( dealId: string, buyOptionId: string): Observable<T> {
    return this.http.put<T>( this.actionUrl.concat('/deal/').concat(dealId).concat('/decrement/buy-option/').concat(buyOptionId), {});
  }

}
